test/state/mpt.cpp

```diff
@@ -0,0 +1,283 @@
+    bytes m_value;
+    std::unique_ptr<MPTNode> m_children[num_children];
+
+    explicit MPTNode(Kind kind, const Path& path = {}, bytes&& value = {}) noexcept
+      : m_kind{kind}, m_path{path}, m_value{std::move(value)}
+    {}
+
+    /// Named constructor for a branch node.
+    static MPTNode branch(size_t idx1, std::unique_ptr<MPTNode> child1, size_t idx2,
+        std::unique_ptr<MPTNode> child2) noexcept
+    {
+        assert(idx1 != idx2);
+        assert(idx1 < num_children);
+        assert(idx2 < num_children);
+        MPTNode node{Kind::branch};
+        node.m_children[idx1] = std::move(child1);
+        node.m_children[idx2] = std::move(child2);
+        return node;
+    }
+
+    /// Named constructor for an extended node.
+    static MPTNode ext(const Path& path, std::unique_ptr<MPTNode> child) noexcept
+    {
+        MPTNode node{Kind::ext, path};
+        node.m_children[0] = std::move(child);
+        return node;
+    }
+
+    /// Finds the position at witch two paths differ.
+    static size_t mismatch(const Path& p1, const Path& p2) noexcept
+    {
+        assert(p1.length <= p2.length);
+        return static_cast<size_t>(
+            std::mismatch(p1.nibbles, p1.nibbles + p1.length, p2.nibbles).first - p1.nibbles);
+    }
+
+public:
+    MPTNode() = default;
+
+    /// Named constructor for a leaf node.
+    static MPTNode leaf(const Path& path, bytes&& value) noexcept
+    {
+        return MPTNode{Kind::leaf, path, std::move(value)};
+    }
+
+    void insert(const Path& path, bytes&& value);
+
+    [[nodiscard]] hash256 hash() const;
+};
+
+void MPTNode::insert(const Path& path, bytes&& value)  // NOLINT(misc-no-recursion)
+{
+    switch (m_kind)
+    {
+    case Kind::branch:
+    {
+        assert(m_path.length == 0);  // Branch has no path.
+
+        const auto idx = path.nibbles[0];
+        auto& child = m_children[idx];
+        if (!child)
+            child = std::make_unique<MPTNode>(leaf(path.tail(1), std::move(value)));
+        else
+            child->insert(path.tail(1), std::move(value));
+        break;
+    }
+
+    case Kind::ext:
+    {
+        assert(m_path.length != 0);  // Ext must have non-empty path.
+
+        const auto mismatch_pos = mismatch(m_path, path);
+
+        if (mismatch_pos == m_path.length)
+        {
+            // Paths match: go into the child.
+            return m_children[0]->insert(path.tail(mismatch_pos), std::move(value));
+        }
+
+        const auto orig_idx = m_path.nibbles[mismatch_pos];
+        const auto new_idx = path.nibbles[mismatch_pos];
+
+        auto br = branch(orig_idx,
+            (mismatch_pos < m_path.length - 1) ?
+                std::make_unique<MPTNode>(
+                    ext(m_path.tail(mismatch_pos + 1), std::move(m_children[0]))) :
+                std::move(m_children[0]),
+            new_idx,
+            std::make_unique<MPTNode>(leaf(path.tail(mismatch_pos + 1), std::move(value))));
+
+        *this = (mismatch_pos == 0) ?
```

Comment:
What if `mismatch_pos == 1`, it will create new extension node with 1-nibble path, is this correct?